### DISCUSSION TITLE
overlay for generated packages

### DISCUSSION THREAD

**Discussion by benneti** [nrdxp/nixdev#365] [Q&A]
Hey, is there a clever way to create an overlay for generated packages, in particular kernelPackages. Right now I only create the overlay for the package of the kernel I use, but I think this is not too elegant.

**Comment by blaggacao** [comment]
Can you paste some code? I unfortunately  have no visual at all at the moment.

**Reply by benneti** [reply]
Ah no problem, right now I have the following in profiles/hardware/tuxedo.nix
``` nix
{ config, lib, pkgs, ... }: {

  boot = {
    extraModulePackages = with config.boot.kernelPackages;
      [
        (tuxedo-keyboard.overrideAttrs (self: rec {
          inherit (pkgs.sources.tuxedo-keyboard) version src;

          makeFlags = self.makeFlags ++ [ "all" ];
          installPhase = self.installPhase + ''
            find src -name "*.ko" -exec mv {} $out/lib/modules/* \;
          '';
        })) # HACK newest version as well as all kernel modules package
      ];
  };
  environment.systemPackages = with pkgs;
    [
      # tuxedo-control-center # FIXME the daemon is broken
      tuxedo-touchpad-switch
    ];
  services.udev.packages = with pkgs; [ tuxedo-touchpad-switch ];
}
```
this achieves what I want in part, because it automatically overwrites tuxedo-keyboard for the currently used kernel.
But I would prefer to have the override of the package in overlays/tuxedo-keyboard.nix . 
Here the problem is that I do not think I can access config, instead I would like to override     `linuxPackages*.tuxedo-keyboard` and let the lazy evaluation handle that I only apply it for the installed one.

Does this clear it up a bit more?

**Reply by blaggacao** [reply]
Is the limitation that you are hitting that overlays are not recursive, that is only work for top level attributes?

**Reply by benneti** [reply]
I do not completely understand how you mean this.
The problem is more like I would like the same overlay for all packages matching a pattern.
Do you mean that the problem is that I would so to speak apply the overlay before the packages are parsed or that I would like to apply an overlay for all available kernels.

But it is not too important it works how I do it good enough...

**Reply by blaggacao** [reply]
I think I don't yet grasp the full picture. It also looks like the severity is limited, that's nice.

---

I was referring to the fact that in a `pkgs` with this attributes:

```nix
{
  a.a = "";
  a.b = "";
}
```

```nix
final: prev: { a.c = ""; }
```

would make `a.a` & `a.b` disappear.

**Reply by benneti** [reply]
ah, yeah that is not the problem I am facing. 
I would like to overwrite a.mypackage, b.mypackage, c.mypackage at the same time where package is always the same (a,b,c are different linux kernels for example)
